### DISCUSSION TITLE
Continuos Deployment Strategy with Fast API having Background Tasks

### DISCUSSION THREAD

**Discussion by jainal09** [fastapi/fastapi#6794] [Questions] [answered]
### First check

* [X] I added a very descriptive title to this issue.
* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [X] I already read and followed all the tutorial in the docs and didn't find an answer.
* [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [X] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.

## ISSUE
So,
I want to know the update strategy for my application which is deployed on a Azure VM and that have background tasks running in it. So, if I use something as Kubernetes although it would take a docker pull and handle all the requests efficiently and will only terminate containers that are not receiving any requests and redirect new requests to the newly updated containers.
But, what if the container which is not receiving any requests may have running some background tasks and the kubernetes will stop it and replace it with new container.

NOTE: I am not bound to use Kubernetes I am just giving example.
In general I want to know How do I update my application with new code with Continuous Deployment or lets say even manually I want to stop the uvicorn app. But How do i know that its not running any background tasks in it.

**Comment by jainal09** [comment]
@tiangolo

**Comment by ycd** [comment]
I didn't understand the question and what you are expecting in here. But seems like you need to use more robust message - job queue instead of "BackgroundTasks" because there is no way to recover state and there is no point holding a state in a container that runs on kubernetes, it is designed to work **stateless** (in most cases).

**Comment by mittalsuraj18** [comment, marked as answer]
The general assumption when having a background task in FastApi is that these are short lived background tasks.
For long lived tasks, something like Celery might be a better option.

To make sure you terminate gracefully, the way you would do it in Kubernetes is by using a `preStop` hooks. 
An example of the yaml looks something like this
```
apiVersion: v1
kind: Pod
metadata:
  name: my-pod
spec:
  containers:
    - name: web
      image: nginx
      ports:
        - name: web
          containerPort: 80
      lifecycle:
        preStop:
          exec:
            command: ["sleep", "30"]
```

you can increase the sleep time here to suit your use cases.
what is happening here is, when a new deployment happens, kubernetes migrates all the traffic to the new deployment and waits 30 seconds before sending `SIGTERM` to the container


The same strategy can be applied to different environments from kubernetes.
Also, you can look at `terminationGracePeriodSeconds` property of the pod too.

Reference: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

If that solves the original problem, then you can close this issue @jainal09  ✔️

But agreed that you probably shouldn't use the integrated background tasks for things that you are not okay with losing. In that case, you should probably use an external system like RQ, Celery, or something else.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by jainal09** [comment]
Thanks for the update.
I then used celery for that project.
Thanks again 👍@tiangolo @mittalsuraj18
